File: cmd/processor/service.go
package main

import (
	"encoding/json"
	"io"
	"regexp"

	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
	extproc_cfg "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/ext_proc/v3alpha"
	extproc "github.com/envoyproxy/go-control-plane/envoy/service/ext_proc/v3alpha"
	envoy_type "github.com/envoyproxy/go-control-plane/envoy/type/v3"
)

var contentTypeJson = regexp.MustCompile("^(application|text)/json(;.*)?$")

type processorService struct{}

func (s *processorService) Process(stream extproc.ExternalProcessor_ProcessServer) error {
	msg, err := stream.Recv()
	if err == io.EOF {
		logger.Debug("Stream closed by proxy")
		return nil
	}
	if err != nil {
		logger.Errorf("Error receiving from stream: %s", err)
		return err
	}

	headers := msg.GetRequestHeaders()
	if headers == nil {
		logger.Warn("Expecting request headers message first")
		// Close stream since there's nothing else we can do
		return nil
	}

	path := getHeaderValue(headers.Headers, ":path")
	logger.Debugf("Received request headers for %s", path)

	switch path {
	case "/echo":
	case "/help":
	case "/hello":
	case "/json":
		// Pass through the basic ones so that the test target works as designed.
		// Just close the stream, which indicates "no more processing"
		return nil
	case "/addHeader":
		return processAddHeader(stream)
	case "/checkJson":
		return processCheckJson(stream, headers)
	case "/notfound":
		return processNotFound(stream)
	}

	// Do nothing for any unknown paths
	return nil
}

// Show how to return an error by returning a 404 in response to the "/notfound"
// path.
func processNotFound(stream extproc.ExternalProcessor_ProcessServer) error {
	// Construct an "immediate" response that will go back to the caller. It will
	// have:
	// * A status code of 404
	// * a content-type header of "text/plain"
	// * a body that says "Not found"
	// * an additional message that may be logged by envoy
	response := &extproc.ProcessingResponse{
		Response: &extproc.ProcessingResponse_ImmediateResponse{
			ImmediateResponse: &extproc.ImmediateResponse{
				Status: &envoy_type.HttpStatus{
					Code: envoy_type.StatusCode_NotFound,
				},
				Headers: &extproc.HeaderMutation{
					SetHeaders: []*core.HeaderValueOption{
						{
							Header: &core.HeaderValue{
								Key:   "content-type",
								Value: "text/plain",
							},
						},
					},
				},
				Body:    "Not found",
				Details: "Requested path was not found",
			},
		},
	}
	// Send the message and return, which closes the stream
	return stream.Send(response)
}

// Show how to add a header to the response.
func processAddHeader(stream extproc.ExternalProcessor_ProcessServer) error {
	// Change the path to "/hello" because that's one of the paths that the target
	// server understands. (Sadly, go syntax and the way that it handles "oneof"
	// messages means a lot of boilerplate here!)
	err := stream.Send(&extproc.ProcessingResponse{
		Response: &extproc.ProcessingResponse_RequestHeaders{
			RequestHeaders: &extproc.HeadersResponse{
				Response: &extproc.CommonResponse{
					HeaderMutation: &extproc.HeaderMutation{
						SetHeaders: []*core.HeaderValueOption{
							{
								Header: &core.HeaderValue{
									Key:   ":path",
									Value: "/hello",
								},
							},
						},
					},
				},
			},
		},
	})
	if err != nil {
		return err
	}

	msg, err := stream.Recv()
	if err != nil {
		return err
	}

	// We have not changed the processing mode, so the next message will be a
	// response headers message, and we should close the stream if it is not.
	responseHeaders := msg.GetResponseHeaders()
	if responseHeaders == nil {
		logger.Error("Expecting response headers as the next message")
		return nil
	}

	// Send back a response that adds a header
	response := &extproc.ProcessingResponse{
		Response: &extproc.ProcessingResponse_ResponseHeaders{
			ResponseHeaders: &extproc.HeadersResponse{
				Response: &extproc.CommonResponse{
					HeaderMutation: &extproc.HeaderMutation{
						SetHeaders: []*core.HeaderValueOption{
							{
								Header: &core.HeaderValue{
									Key:   "x-external-processor-status",
									Value: "We were here",
								},
							},
						},
					},
				},
			},
		},
	}
	return stream.Send(response)
}

// This is a more sophisticated example that does a few things:
// 1) It sets a request header to rewrite the path
// 2) It checks the content-type header to see if it is JSON (or skips
//    this check if there is no content
// 3) If it is JSON, it changes the processing mode to get the request body
// 4) If the content is JSON, validate the request body as JSON
// 5) Return an error if the validation fails
// 6) Or, add a response header to reflect the request validation status
//
// Among other things, this example shows the ablility of an external processor
// to decide how much of the request and response it needs to process based
// on input.
func processCheckJson(stream extproc.ExternalProcessor_ProcessServer,
	requestHeaders *extproc.HttpHeaders) error {
	// Set the path to "/echo" to get the right functionality on the target.
	requestHeadersResponse := &extproc.ProcessingResponse{
		Response: &extproc.ProcessingResponse_RequestHeaders{
			RequestHeaders: &extproc.HeadersResponse{
				Response: &extproc.CommonResponse{
					HeaderMutation: &extproc.HeaderMutation{
						SetHeaders: []*core.HeaderValueOption{
							{
								Header: &core.HeaderValue{
									Key:   ":path",
									Value: "/echo",
								},
							},
						},
					},
				},
			},
		},
	}

	// If the content-type header looks like JSON, then ask for the request body
	contentType := getHeaderValue(requestHeaders.Headers, "content-type")
	logger.Debugf("Checking content-type %s to see if it is JSON", contentType)
	contentIsJson := !requestHeaders.EndOfStream && contentTypeJson.MatchString(contentType)
	if contentIsJson {
		requestHeadersResponse.ModeOverride = &extproc_cfg.ProcessingMode{
			RequestBodyMode: extproc_cfg.ProcessingMode_BUFFERED,
		}
	}

	err := stream.Send(requestHeadersResponse)
	if err != nil {
		return err
	}

	var msg *extproc.ProcessingRequest
	var jsonStatus string

	if contentIsJson {
		msg, err = stream.Recv()
		if err != nil {
			return err
		}

		requestBody := msg.GetRequestBody()
		if requestBody == nil {
			logger.Error("Expected request body to be sent next")
			return nil
		}

		requestBodyResponse := &extproc.ProcessingResponse{}

		if json.Valid(requestBody.Body) {
			// Nothing to do, so return an empty response
			requestBodyResponse.Response = &extproc.ProcessingResponse_RequestBody{}

		} else {
			requestBodyResponse.Response = &extproc.ProcessingResponse_ImmediateResponse{
				ImmediateResponse: &extproc.ImmediateResponse{
					Status: &envoy_type.HttpStatus{
						Code: envoy_type.StatusCode_BadRequest,
					},
					Headers: &extproc.HeaderMutation{
						SetHeaders: []*core.HeaderValueOption{
							{
								Header: &core.HeaderValue{
									Key:   "content-type",
									Value: "text/plain",
								},
							},
						},
					},
					Body:    "Invalid JSON",
					Details: "Request body was not valid JSON",
				},
			}
			// Send the error response and end processing now
			return stream.Send(requestBodyResponse)
		}

		// Send the body response and wait for the next message
		err = stream.Send(requestBodyResponse)
		if err != nil {
			return err
		}
		jsonStatus = "Body is valid JSON"

	} else {
		jsonStatus = "Body is not JSON"
	}

	msg, err = stream.Recv()
	if err != nil {
		return err
	}

	responseHeaders := msg.GetResponseHeaders()
	if responseHeaders == nil {
		logger.Error("Expecting response headers as the next message")
		return nil
	}

	responseHeadersResponse := &extproc.ProcessingResponse{
		Response: &extproc.ProcessingResponse_ResponseHeaders{
			ResponseHeaders: &extproc.HeadersResponse{
				Response: &extproc.CommonResponse{
					HeaderMutation: &extproc.HeaderMutation{
						SetHeaders: []*core.HeaderValueOption{
							{
								Header: &core.HeaderValue{
									Key:   "x-json-status",
									Value: jsonStatus,
								},
							},
						},
					},
				},
			},
		},
	}
	return stream.Send(responseHeadersResponse)
}

// getHeaderValue returns the value of the first HTTP header in the map that matches.
// We don't expect that we will need to look up many headers, so simply do a linear search. If
// we needed to query many more headers, we'd turn the header map into a "map[string]string",
// or a "map[string][]string" if we wanted to handle multi-value headers.
// Return the empty string if the header is not found.
func getHeaderValue(headers *core.HeaderMap, name string) string {
	for _, h := range headers.Headers {
		if h.Key == name {
			return h.Value
		}
	}
	return ""
}

File: rust/src/service.rs
use {
  crate::base64::Encoder,
  envoy_control_plane::envoy::{
    config::core::v3::{HeaderMap, HeaderValue, HeaderValueOption},
    extensions::filters::http::ext_proc::v3::{processing_mode, ProcessingMode},
    r#type::v3::HttpStatus,
    service::ext_proc::v3::{
      body_mutation, external_processor_server::ExternalProcessor, processing_request,
      processing_response, BodyMutation, BodyResponse, CommonResponse, HeaderMutation,
      HeadersResponse, HttpBody, HttpHeaders, ImmediateResponse, ProcessingRequest,
      ProcessingResponse,
    },
  },
  futures::{channel::mpsc::UnboundedSender, SinkExt, Stream},
  std::{pin::Pin, str},
  tonic::{Request, Response, Status, Streaming},
};

type ExternalProcessorStream =
  Pin<Box<dyn Stream<Item = Result<ProcessingResponse, Status>> + Send>>;

pub struct ExampleProcessor {}

#[tonic::async_trait]
impl ExternalProcessor for ExampleProcessor {
  type ProcessStream = ExternalProcessorStream;

  async fn process(
    &self,
    request: Request<Streaming<ProcessingRequest>>,
  ) -> Result<Response<ExternalProcessorStream>, Status> {
    let mut stream = request.into_inner();
    if let Some(req_headers) = get_request_headers(&mut stream).await {
      let (sender, receiver) = futures::channel::mpsc::unbounded();
      let path = get_header_value(&req_headers.headers, ":path");
      match path {
        Some("/notfound") => handle_not_found(sender)?,
        Some("/addHeader") => {
          tokio::task::spawn(async move {
            handle_add_header(sender, stream).await;
          });
        }
        Some("/checkJson") => {
          tokio::task::spawn(async move {
            handle_check_json(&req_headers, sender, stream).await;
          });
        }
        Some("/echoencode") => {
          tokio::task::spawn(async move {
            handle_echo_encode(sender, stream).await;
          });
        }
        _ => sender.close_channel(),
      }
      return Ok(Response::new(Box::pin(receiver)));
    }
    // By default, just close the stream.
    Ok(Response::new(Box::pin(futures::stream::empty())))
  }
}

// Handle a not found by immediately writing to the channel and letting
// it close.
fn handle_not_found(
  sender: UnboundedSender<Result<ProcessingResponse, Status>>,
) -> Result<(), Status> {
  if sender
    .unbounded_send(Ok(new_error(404, "not found")))
    .is_err()
  {
    return Err(Status::internal("stream error"));
  }
  Ok(())
}

// Add a header to the response.
async fn handle_add_header(
  mut sender: UnboundedSender<Result<ProcessingResponse, Status>>,
  mut stream: Streaming<ProcessingRequest>,
) {
  // Send back a response that changes the request header for the HTTP target.
  let mut req_headers_cr = CommonResponse::default();
  add_set_header(&mut req_headers_cr, ":path", "/hello");
  let req_headers_resp = ProcessingResponse {
    response: Some(processing_response::Response::RequestHeaders(
      HeadersResponse {
        response: Some(req_headers_cr),
      },
    )),
    ..Default::default()
  };
  sender.send(Ok(req_headers_resp)).await.ok();

  if get_response_headers(&mut stream).await.is_some() {
    let mut resp_headers_cr = CommonResponse::default();
    add_set_header(
      &mut resp_headers_cr,
      "x-external-processor-status",
      "Powered by Rust!",
    );
    let resp_headers_resp = ProcessingResponse {
      response: Some(processing_response::Response::ResponseHeaders(
        HeadersResponse {
          response: Some(resp_headers_cr),
        },
      )),
      ..Default::default()
    };
    sender.send(Ok(resp_headers_resp)).await.ok();
  }
  // Fall through if we get the wrong message.
}

// Check that the request body is JSON, and if so, reject the request
// if it is invalid and add a header to the response otherwise.
async fn handle_check_json(
  request_headers: &HttpHeaders,
  mut sender: UnboundedSender<Result<ProcessingResponse, Status>>,
  mut stream: Streaming<ProcessingRequest>,
) {
  let is_json = matches!(
    get_header_value(&request_headers.headers, "content-type"),
    Some("application/json")
  );
  let mut req_headers_cr = CommonResponse::default();
  add_set_header(&mut req_headers_cr, ":path", "/echo");
  let mut req_headers_resp = ProcessingResponse {
    response: Some(processing_response::Response::RequestHeaders(
      HeadersResponse {
        response: Some(req_headers_cr),
      },
    )),
    ..Default::default()
  };
  if is_json {
    // Switch to a mode in which we get the body only if it's JSON.
    req_headers_resp.mode_override = Some(ProcessingMode {
      request_body_mode: processing_mode::BodySendMode::Buffered as i32,
      ..Default::default()
    });
  }
  sender.send(Ok(req_headers_resp)).await.ok();

  let mut json_status = "Not JSON";

  if is_json {
    if let Some(request_body) = get_request_body(&mut stream).await {
      if let Ok(body_str) = str::from_utf8(&request_body.body) {
        match json::parse(body_str) {
          Ok(_) => json_status = "Valid JSON",
          Err(_) => {
            sender.send(Ok(new_error(400, "Invalid JSON"))).await.ok();
            return;
          }
        }
      }
    }
    let req_body_response = ProcessingResponse {
      response: Some(processing_response::Response::RequestBody(
        BodyResponse::default(),
      )),
      ..Default::default()
    };
    sender.send(Ok(req_body_response)).await.ok();
  }

  if get_response_headers(&mut stream).await.is_some() {
    let mut resp_headers_cr = CommonResponse::default();
    add_set_header(&mut resp_headers_cr, "x-json-status", json_status);
    let resp_headers_resp = ProcessingResponse {
      response: Some(processing_response::Response::ResponseHeaders(
        HeadersResponse {
          response: Some(resp_headers_cr),
        },
      )),
      ..Default::default()
    };
    sender.send(Ok(resp_headers_resp)).await.ok();
  }
  // Fall through if we get the wrong message.
}

// Encode the response in a streaming way into base64.
async fn handle_echo_encode(
  mut sender: UnboundedSender<Result<ProcessingResponse, Status>>,
  mut stream: Streaming<ProcessingRequest>,
) {
  // Send back a response that changes the request URL for the HTTP target.
  let mut req_headers_cr = CommonResponse::default();
  add_set_header(&mut req_headers_cr, ":path", "/echo");
  let req_headers_resp = ProcessingResponse {
    response: Some(processing_response::Response::RequestHeaders(
      HeadersResponse {
        response: Some(req_headers_cr),
      },
    )),
    mode_override: Some(ProcessingMode {
      response_body_mode: processing_mode::BodySendMode::Streamed as i32,
      ..Default::default()
    }),
    ..Default::default()
  };
  sender.send(Ok(req_headers_resp)).await.ok();

  let mut encoder = Encoder::new();

  // Loop to process messages, because we act on both response headers
  // and also on each chunk of the body.
  while let Ok(Some(next_msg)) = stream.message().await {
    match next_msg.request {
      Some(processing_request::Request::ResponseHeaders(_)) => {
        let resp_headers_resp = ProcessingResponse {
          // Be sure to change the path so that the HTTP target works,
          // and clear content-length since it will change as we encode.
          response: Some(processing_response::Response::ResponseHeaders(
            HeadersResponse {
              response: Some(CommonResponse {
                header_mutation: Some(HeaderMutation {
                  set_headers: vec![HeaderValueOption {
                    header: Some(HeaderValue {
                      key: ":path".into(),
                      value: "/echo".into(),
                    }),
                    ..Default::default()
                  }],
                  remove_headers: vec!["content-length".into()],
                }),
                ..Default::default()
              }),
            },
          )),
          ..Default::default()
        };
        sender.send(Ok(resp_headers_resp)).await.ok();
      }
      Some(processing_request::Request::ResponseBody(chunk)) => {
        let new_body = encoder.encode(&chunk.body, chunk.end_of_stream);
        let resp_body_resp = ProcessingResponse {
          response: Some(processing_response::Response::ResponseBody(BodyResponse {
            response: Some(CommonResponse {
              body_mutation: Some(BodyMutation {
                mutation: Some(body_mutation::Mutation::Body(new_body.into())),
              }),
              ..Default::default()
            }),
          })),
          ..Default::default()
        };
        sender.send(Ok(resp_body_resp)).await.ok();
      }
      _ => {}
    }
  }
}

async fn get_request_headers(stream: &mut Streaming<ProcessingRequest>) -> Option<HttpHeaders> {
  if let Ok(Some(next_msg)) = stream.message().await {
    if let Some(processing_request::Request::RequestHeaders(hdrs)) = next_msg.request {
      return Some(hdrs);
    }
  }
  None
}

async fn get_request_body(stream: &mut Streaming<ProcessingRequest>) -> Option<HttpBody> {
  if let Ok(Some(next_msg)) = stream.message().await {
    if let Some(processing_request::Request::RequestBody(hdrs)) = next_msg.request {
      return Some(hdrs);
    }
  }
  None
}

async fn get_response_headers(stream: &mut Streaming<ProcessingRequest>) -> Option<HttpHeaders> {
  if let Ok(Some(next_msg)) = stream.message().await {
    if let Some(processing_request::Request::ResponseHeaders(hdrs)) = next_msg.request {
      return Some(hdrs);
    }
  }
  None
}

fn new_error(status: i32, msg: &str) -> ProcessingResponse {
  ProcessingResponse {
    response: Some(processing_response::Response::ImmediateResponse(
      ImmediateResponse {
        status: Some(HttpStatus { code: status }),
        details: msg.into(),
        ..Default::default()
      },
    )),
    ..Default::default()
  }
}

fn get_header_value<'a>(header_map: &'a Option<HeaderMap>, name: &str) -> Option<&'a str> {
  match header_map {
    Some(headers) => {
      for header in &headers.headers {
        if header.key == name {
          return Some(&header.value);
        }
      }
      None
    }
    None => None,
  }
}

fn add_set_header(cr: &mut CommonResponse, key: &str, value: &str) {
  let new_header = HeaderValueOption {
    header: Some(HeaderValue {
      key: key.into(),
      value: value.into(),
    }),
    ..Default::default()
  };
  match &mut cr.header_mutation {
    Some(hm) => hm.set_headers.push(new_header),
    None => {
      let mut new_hm = HeaderMutation::default();
      new_hm.set_headers.push(new_header);
      cr.header_mutation = Some(new_hm);
    }
  }
}
